Failing CuPy tests

Since the introduction of https://github.com/dask/dask/pull/6738, a few CuPy tests have broken.

- 1 `bincount` test fails after #7183 ;
- 16 `tri*` tests fail after #6997 .

We're discussing and trying to allocate GPUs for Dask PRs in the near future, so I'll probably wait for that to work on fixing the tests above.

@GenevieveBuckley thanks for pinging me, but unfortunately I still see the same failure with #7391 : 

```python
======================================================= FAILURES =======================================================
____________________________________________________ test_bincount _____________________________________________________

    @pytest.mark.skipif(
        not IS_NEP18_ACTIVE or cupy.__version__ < LooseVersion("6.4.0"),
        reason="NEP-18 support is not available in NumPy or CuPy older than "
        "6.4.0 (requires https://github.com/cupy/cupy/pull/2418)",
    )
    def test_bincount():
        x = cupy.array([2, 1, 5, 2, 1])
        d = da.from_array(x, chunks=2, asarray=False)
        e = da.bincount(d, minlength=6)
>       assert_eq(e, np.bincount(x, minlength=6))

dask/array/tests/test_cupy.py:934:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = array([0, 2, 2, 0, 0, 1]), b = array([0, 2, 2, 0, 0, 1]), check_shape = True, check_graph = True, check_meta = True
check_chunks = True, kwargs = {}
a_original = dask.array<_bincount_agg-aggregate, shape=(6,), dtype=int64, chunksize=(6,), chunktype=numpy.ndarray>
b_original = array([0, 2, 2, 0, 0, 1]), adt = dtype('int64'), a_meta = array([], dtype=int64)
a_computed = array([0, 2, 2, 0, 0, 1]), bdt = dtype('int64'), b_meta = None

    def assert_eq(
        a,
        b,
        check_shape=True,
        check_graph=True,
        check_meta=True,
        check_chunks=True,
        **kwargs,
    ):
        a_original = a
        b_original = b

        a, adt, a_meta, a_computed = _get_dt_meta_computed(
            a, check_shape=check_shape, check_graph=check_graph, check_chunks=check_chunks
        )
        b, bdt, b_meta, b_computed = _get_dt_meta_computed(
            b, check_shape=check_shape, check_graph=check_graph, check_chunks=check_chunks
        )

        if str(adt) != str(bdt):
            # Ignore check for matching length of flexible dtypes, since Array._meta
            # can't encode that information
            if adt.type == bdt.type and not (adt.type == np.bytes_ or adt.type == np.str_):
                diff = difflib.ndiff(str(adt).splitlines(), str(bdt).splitlines())
                raise AssertionError(
                    "string repr are different" + os.linesep + os.linesep.join(diff)
                )

        try:
            assert (
                a.shape == b.shape
            ), f"a and b have different shapes (a: {a.shape}, b: {b.shape})"
            if check_meta:
                if hasattr(a, "_meta") and hasattr(b, "_meta"):
                    assert_eq(a._meta, b._meta)
                if hasattr(a_original, "_meta"):
                    msg = (
                        f"compute()-ing 'a' changes its number of dimensions "
                        f"(before: {a_original._meta.ndim}, after: {a.ndim})"
                    )
                    assert a_original._meta.ndim == a.ndim, msg
                    if a_meta is not None:
                        msg = (
                            f"compute()-ing 'a' changes its type "
                            f"(before: {type(a_original._meta)}, after: {type(a_meta)})"
                        )
                        assert type(a_original._meta) == type(a_meta), msg
                        if not (np.isscalar(a_meta) or np.isscalar(a_computed)):
                            msg = (
                                f"compute()-ing 'a' results in a different type than implied by its metadata "
                                f"(meta: {type(a_meta)}, computed: {type(a_computed)})"
                            )
>                           assert type(a_meta) == type(a_computed), msg
E                           AssertionError: compute()-ing 'a' results in a different type than implied by its metadata (meta: <class 'numpy.ndarray'>, computed: <class 'cupy.core.core.ndarray'>)
E                           assert <class 'numpy.ndarray'> == <class 'cupy....core.ndarray'>
E                             +<class 'numpy.ndarray'>
E                             -<class 'cupy.core.core.ndarray'>

dask/array/utils.py:320: AssertionError
```

--- a/dask/array/chunk.py
+++ b/dask/array/chunk.py
@@ -261,6 +261,7 @@ def argtopk_aggregate(a_plus_idx, k, axis, keepdims):
 
 def arange(start, stop, step, length, dtype, like=None):
     from .utils import arange_safe
+
     res = arange_safe(start, stop, step, dtype, like=like)
     return res[:-1] if len(res) > length else res
 

--- a/dask/array/chunk.py
+++ b/dask/array/chunk.py
@@ -259,8 +259,9 @@ def argtopk_aggregate(a_plus_idx, k, axis, keepdims):
     ]
 
 
-def arange(start, stop, step, length, dtype):
-    res = np.arange(start, stop, step, dtype)
+def arange(start, stop, step, length, dtype, like=None):
+    from .utils import arange_safe
+    res = arange_safe(start, stop, step, dtype, like=like)
     return res[:-1] if len(res) > length else res
 
 

--- a/dask/array/creation.py
+++ b/dask/array/creation.py
@@ -378,6 +378,9 @@ def arange(*args, **kwargs):
 
     num = int(max(np.ceil((stop - start) / step), 0))
 
+    like = kwargs.pop("like", None)
+    meta = meta_from_array(like) if like is not None else None
+
     dtype = kwargs.pop("dtype", None)
     if dtype is None:
         dtype = np.arange(start, stop, step * num if num else step).dtype
@@ -394,11 +397,11 @@ def arange(*args, **kwargs):
     for i, bs in enumerate(chunks[0]):
         blockstart = start + (elem_count * step)
         blockstop = start + ((elem_count + bs) * step)
-        task = (chunk.arange, blockstart, blockstop, step, bs, dtype)
+        task = (partial(chunk.arange, like=like), blockstart, blockstop, step, bs, dtype)
         dsk[(name, i)] = task
         elem_count += bs
 
-    return Array(dsk, name, chunks, dtype=dtype)
+    return Array(dsk, name, chunks, dtype=dtype, meta=meta)
 
 
 @derived_from(np)
@@ -677,7 +680,7 @@ def diagonal(a, offset=0, axis1=0, axis2=1):
 
 
 @derived_from(np)
-def tri(N, M=None, k=0, dtype=float, chunks="auto"):
+def tri(N, M=None, k=0, dtype=float, chunks="auto", like=None):
     _min_int = np.lib.twodim_base._min_int
 
     if M is None:
@@ -685,9 +688,9 @@ def tri(N, M=None, k=0, dtype=float, chunks="auto"):
 
     chunks = normalize_chunks(chunks, shape=(N, M), dtype=dtype)
 
-    m = greater_equal.outer(
-        arange(N, chunks=chunks[0][0], dtype=_min_int(0, N)),
-        arange(-k, M - k, chunks=chunks[1][0], dtype=_min_int(-k, M - k)),
+    m = greater_equal(
+        arange(N, chunks=chunks[0][0], dtype=_min_int(0, N), like=like).reshape(1, N).T,
+        arange(-k, M - k, chunks=chunks[1][0], dtype=_min_int(-k, M - k), like=like),
     )
 
     # Avoid making a copy if the requested type is already bool

--- a/dask/array/creation.py
+++ b/dask/array/creation.py
@@ -397,7 +397,14 @@ def arange(*args, **kwargs):
     for i, bs in enumerate(chunks[0]):
         blockstart = start + (elem_count * step)
         blockstop = start + ((elem_count + bs) * step)
-        task = (partial(chunk.arange, like=like), blockstart, blockstop, step, bs, dtype)
+        task = (
+            partial(chunk.arange, like=like),
+            blockstart,
+            blockstop,
+            step,
+            bs,
+            dtype,
+        )
         dsk[(name, i)] = task
         elem_count += bs
 

--- a/dask/array/routines.py
+++ b/dask/array/routines.py
@@ -1721,18 +1721,20 @@ def average(a, axis=None, weights=None, returned=False):
 
 @derived_from(np)
 def tril(m, k=0):
-    m = asarray(m)
-    mask = tri(*m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:])
+    from .utils import asarray_safe
+    m = asarray_safe(m, like=m)
+    mask = tri(*m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m))
 
-    return where(mask, m, zeros(1, dtype=m.dtype))
+    return where(mask, m, zeros_like_safe(m, shape=(1, )))
 
 
 @derived_from(np)
 def triu(m, k=0):
-    m = asarray(m)
-    mask = tri(*m.shape[-2:], k=k - 1, dtype=bool, chunks=m.chunks[-2:])
+    from .utils import asarray_safe
+    m = asarray_safe(m, like=m)
+    mask = tri(*m.shape[-2:], k=k - 1, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m))
 
-    return where(mask, zeros(1, dtype=m.dtype), m)
+    return where(mask, zeros_like_safe(m, shape=(1, )), m)
 
 
 @derived_from(np)

--- a/dask/array/routines.py
+++ b/dask/array/routines.py
@@ -23,6 +23,7 @@ from .utils import (
     meta_from_array,
     zeros_like_safe,
     array_safe,
+    asarray_safe,
 )
 from .wrap import ones
 from .ufunc import multiply, sqrt
@@ -684,8 +685,6 @@ def bincount(x, weights=None, minlength=0, split_every=None):
 
 @derived_from(np)
 def digitize(a, bins, right=False):
-    from .utils import asarray_safe
-
     bins = asarray_safe(bins, like=meta_from_array(a))
     dtype = np.digitize(asarray_safe([0], like=bins), bins, right=False).dtype
     return a.map_blocks(np.digitize, dtype=dtype, bins=bins, right=right)
@@ -1721,8 +1720,6 @@ def average(a, axis=None, weights=None, returned=False):
 
 @derived_from(np)
 def tril(m, k=0):
-    from .utils import asarray_safe
-
     m = asarray_safe(m, like=m)
     mask = tri(
         *m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m)
@@ -1733,8 +1730,6 @@ def tril(m, k=0):
 
 @derived_from(np)
 def triu(m, k=0):
-    from .utils import asarray_safe
-
     m = asarray_safe(m, like=m)
     mask = tri(
         *m.shape[-2:],

--- a/dask/array/routines.py
+++ b/dask/array/routines.py
@@ -16,7 +16,7 @@ from ..delayed import unpack_collections, Delayed
 from ..highlevelgraph import HighLevelGraph
 from ..utils import funcname, derived_from, is_arraylike
 from . import chunk
-from .creation import arange, diag, empty, indices, tri, zeros
+from .creation import arange, diag, empty, indices, tri
 from .utils import (
     safe_wraps,
     validate_axis,
@@ -1722,19 +1722,29 @@ def average(a, axis=None, weights=None, returned=False):
 @derived_from(np)
 def tril(m, k=0):
     from .utils import asarray_safe
-    m = asarray_safe(m, like=m)
-    mask = tri(*m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m))
 
-    return where(mask, m, zeros_like_safe(m, shape=(1, )))
+    m = asarray_safe(m, like=m)
+    mask = tri(
+        *m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m)
+    )
+
+    return where(mask, m, zeros_like_safe(m, shape=(1,)))
 
 
 @derived_from(np)
 def triu(m, k=0):
     from .utils import asarray_safe
-    m = asarray_safe(m, like=m)
-    mask = tri(*m.shape[-2:], k=k - 1, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m))
 
-    return where(mask, zeros_like_safe(m, shape=(1, )), m)
+    m = asarray_safe(m, like=m)
+    mask = tri(
+        *m.shape[-2:],
+        k=k - 1,
+        dtype=bool,
+        chunks=m.chunks[-2:],
+        like=meta_from_array(m),
+    )
+
+    return where(mask, zeros_like_safe(m, shape=(1,)), m)
 
 
 @derived_from(np)

--- a/dask/array/utils.py
+++ b/dask/array/utils.py
@@ -426,6 +426,21 @@ def zeros_like_safe(a, shape, **kwargs):
         return np.zeros(shape, **kwargs)
 
 
+def arange_safe(*args, like, **kwargs):
+    """
+    Use the `like=` from `np.arange` to create a new array dispatching
+    to the downstream library. If that fails, falls back to the
+    default NumPy behavior, resulting in a `numpy.ndarray`.
+    """
+    if like is None:
+        return np.arange(*args, **kwargs)
+    else:
+        try:
+            return np.arange(*args, like=meta_from_array(like), **kwargs)
+        except TypeError:
+            return np.arange(*args, **kwargs)
+
+
 def _array_like_safe(np_func, da_func, a, like, **kwargs):
     if like is a and hasattr(a, "__array_function__"):
         return a

